File: naru_agent/knowledge/contextualizer.py
"""Contextual Retrieval: enrich chunks with situational context at ingest time."""

from __future__ import annotations


class ChunkContextualizer:
    """在 ingestion 時用 LLM 為每個 chunk 補充定位上下文。

    Anthropic 實測：搭配 BM25 降低 49% 檢索失敗率，搭配 reranker 降低 67%。
    成本：只在建庫時執行一次，查詢時零額外開銷。

    Args:
        model: LiteLLM 相容的模型 ID（預設輕量模型）
        api_key: 模型 API key（可從環境變數讀取）
    """

    PROMPT_TEMPLATE = """\
<document>
{document}
</document>

以下是要定位的 chunk：
<chunk>
{chunk}
</chunk>

請用一到三句話說明這個 chunk 在整份文件中的位置和背景脈絡，\
目的是改善向量搜索的準確度。只輸出這段說明，不要其他內容。"""

    def __init__(
        self,
        model: str = "gemini/gemini-2.5-flash-lite",
        api_key: str | None = None,
    ) -> None:
        self._model = model
        self._api_key = api_key

    def contextualize(self, chunk: str, full_document: str) -> str:
        """為單一 chunk 生成定位上下文，回傳「上下文 + 原始 chunk」的合併文字。"""
        import litellm

        prompt = self.PROMPT_TEMPLATE.format(document=full_document, chunk=chunk)
        kwargs: dict = {
            "model": self._model,
            "messages": [{"role": "user", "content": prompt}],
        }
        if self._api_key:
            kwargs["api_key"] = self._api_key
        response = litellm.completion(**kwargs)
        context = response.choices[0].message.content.strip()
        return f"{context}\n{chunk}"

File: js/src/knowledge/contextualizer.ts
/**
 * Contextual Retrieval: enrich chunks with situational context at ingest time.
 *
 * Anthropic research shows this reduces retrieval failure by 49-67%.
 * Cost: runs once at ingestion; zero overhead at query time.
 */

import { generateText, type LanguageModel } from "ai";

export interface ChunkContextualizerConfig {
  model: LanguageModel;
}

const PROMPT_TEMPLATE = `<document>
{document}
</document>

以下是要定位的 chunk：
<chunk>
{chunk}
</chunk>

請用一到三句話說明這個 chunk 在整份文件中的位置和背景脈絡，\
目的是改善向量搜索的準確度。只輸出這段說明，不要其他內容。`;

export class ChunkContextualizer {
  private readonly model: LanguageModel;

  constructor(config: ChunkContextualizerConfig) {
    this.model = config.model;
  }

  async contextualize(chunk: string, fullDocument: string): Promise<string> {
    const prompt = PROMPT_TEMPLATE
      .replace("{document}", fullDocument.replaceAll("{chunk}", "\\{chunk\\}"))
      .replace("{chunk}", chunk);

    const { text } = await generateText({
      model: this.model,
      messages: [{ role: "user", content: prompt }],
    });

    return `${text.trim()}\n${chunk}`;
  }

  async contextualizeMany(
    chunks: string[],
    fullDocument: string,
    concurrency = 5,
  ): Promise<string[]> {
    const results: string[] = new Array(chunks.length);
    let nextIndex = 0;

    const worker = async () => {
      while (nextIndex < chunks.length) {
        const i = nextIndex++;
        results[i] = await this.contextualize(chunks[i], fullDocument);
      }
    };

    await Promise.all(
      Array.from(
        { length: Math.min(concurrency, chunks.length) },
        () => worker(),
      ),
    );
    return results;
  }
}
